synapse/lib/stormlib/model.py

```diff
@@ -1268,3 +1319,357 @@ async def _riskHasVulnToVulnerable(self, n, nodata=False):
+
+                        { for ($tagname, $tagvalu) in $edit.tagprops {
+                            for ($name, $valu) in $tagvalu {
+                                [ +#$tagname:$name = $valu.0 ]
+                            }
+                        }}
+
+                        { for ($name, $valu) in $edit.data {
+                            $node.data.set($name, $valu)
+                        }}
+                    }
+                }
+            }
+
+            for $offs in $oldnode.offsets.edges {
+                $edges = $edgeq.get($offs, cull=(false), wait=(false))
+                if (not $edges) {
+                    $lib.warn(`Queue node {$offset} missing edges offset {$offs}. Node may be incomplete after restoration.`)
+                    continue
+                }
+
+                $edges = $edges.1
+
+                for $edge in $edges {
+                    if (not $_checkValidLayer($edge.view, $edge.layer)) {
+                        $lib.warn(`Could not apply edge in view {$edge.view}, skipping.`)
+                        continue
+                    }
+
+                    view.exec $edge.view {
+                        *$oldnode.form=$newval
+
+                        switch $edge.direction {
+                            "n1": {
+                                [ +($edge.verb)> { yield $edge.node } ]
+                            }
+
+                            "n2": {
+                                [ <($edge.verb)+ { yield $edge.node } ]
+                            }
+                        }
+                    }
+                }
+            }
+
+            for $offs in $oldnode.offsets.refs {
+                $refs = $refsq.get($offs, cull=(false), wait=(false))
+                if (not $refs) {
+                    $lib.warn(`Queue node {$offset} missing refs offset {$offs}. Node may be incomplete after restoration.`)
+                    continue
+                }
+
+                $refs = $refs.1
+
+                for $ref in $refs {
+                    if (not $_checkValidLayer($ref.view, $ref.layer)) {
+                        $lib.warn(`Could not restore reference in view {$ref.view}, skipping.`)
+                        continue
+                    }
+
+                    ($form, $prop, $type, $isarray) = $ref.refinfo
+
+                    view.exec $ref.view {
+
+                        yield $ref.iden
+
+                        empty {
+                            $lib.warn(`Skipping invalid reference: {$offs} ({$form}:{$prop}).`)
+                            continue
+                        }
+
+                        if ($node.form() != $form) {
+                            $lib.warn(`Skipping invalid reference: expected {$form}, got {$node.form()}.`)
+                            continue
+                        }
+
+                        { -:$prop
+                            if $isarray {
+                                [ :$prop += { *$oldnode.form=$newval } ]
```

Comment:
If the prop is an array do we want to add it even if the prop already has a value since we'll be appending rather than smashing the existing value?

Reply:
Actually, there's a bug here. Fixed to correctly up the array or the unset prop.